### ISSUE TITLE
[RFC] Softmax quantization problematic

### ISSUE THREAD

**Issue by tlepley-cadence** [pytorch/glow#1512]
When adding the support for Softmax quantization, I ran into an issue due to the fact that
- the Softmax node is declared as propagating  the input type to the output
- The Softmax instruction checks that the input type is equal to the output type
- In Glow, the scale/offset are part of the tensor type

With our model, the quantization profile scale is substantially different between the input and the output (input scale = 0.245214, out scale = 0.00682523). Having the input scale propagated to the output makes the quantized Softmax generate only 0 in many cases.

Here, this is really a _scale_ issue and adding a `RescaleQuantized `node after `Softmax` to match the profile scale, like it's done for `MaxPooling` for instance, would not help (requantizing 0 will still gives 0)

I'd like to have your opinion for the best approach to take. Maybe the main source of the problem is that the scale/offset information is in `Type`. If it was moved out of `Type`, I think it would solve many problem swe are going to encounter as we add more nodes that support quantization. 

In the meantime, the workaround I implemented was to change the declaration of both `Softmax` node and instruction (see  PR #1511).
Summary:
- declare `.addResultFromCtorArg()` for the node declaration
- remove  `.inplaceOperand({"Dest", "Src"})` and  `.autoVerify(VerifyKind::SameType, {"Dest", "Src"})` for the instruction declaration
- I also had to add a more relaxed `Type` equality checking method. Since scale/offset are part of the `Type` objects, the `isEqual` methods is in some situation a bit too strict. Sometimes, we just want to check that the shape and the element types are equal, not the scale/offset.

**Comment by tlepley-cadence** [comment]
Closing since the PR was merged.
